### PR TITLE
feat(speculative): add Kimi K3 DSpark training

### PR DESCRIPTION
**Pull request by khazic**
## Summary

- add a dense Kimi K3 MLA backbone for DSpark draft training
- load the frozen K3 target through expert parallelism and FSDP2
- support both online target capture and distributed offline precompute
- add a cluster-ready Kimi K3 recipe and training guide
- cover draft configuration, registry dispatch, forward and backward behavior, and target construction

## Design

The draft preserves K3 Q-LoRA, compressed KV projections, output gating, RMSNorm, and SiTU. It intentionally removes KDA, routed and shared experts, and attention-residual mixing from the trainable draft.

This was originally stacked on #3259, which is now merged, so this branch is self-contained and can be reviewed on its own.

## Validation

- `/simplify`
- `ruff check` on changed Python files
- `ruff format --check` on changed Python files
- `1355 passed, 12 skipped` across `tests/unit_tests/speculative` and `tests/unit_tests/recipes/llm`, covering Kimi K3 DSpark, registry, DeepSeek V4 draft, target capture, offline cache, and the DSpark validation aggregation
- measured patch coverage locally at 91.1% (163 of 179 added statements in `nemo_automodel/`); `draft_kimi_k3.py` is at 100%

### CODE REVIEW

**Comment by khazic** [comment]
@HuiyingLi could I ask you to run this one on GPU when you have a slot? I have no Kimi K3 target weights on my side, so everything below is CPU-verified only and the accuracy numbers are exactly what I cannot produce here.

Two commits since the first push.

**1. `refactor(speculative): tighten Kimi K3 DSpark draft and its docs`**

* `_expand_key_value_groups` was dead code. `kv_b_proj` emits `num_heads * (nope + v)` columns and the caller views that with `num_key_value_heads`, so the module only ever runs when the two are equal and the expansion branch is unreachable. It is now an invariant assert in `__init__`, and the view uses `num_heads` directly, matching GLM-5.2 MLA.
* `build_kimi_k3_draft_config` was missing the three `model_args` preconditions its four sibling builders all check, so a YAML without `markov_head_type` raised a bare `AttributeError` from inside the builder instead of a config error.
* Error text and docstrings still enumerated four model families, and the example YAML plus README carried site-specific paths and an RDMA workaround that belong to one cluster rather than to Kimi K3.

**2. `feat(speculative): report DSpark acceptance diagnostics in validation`**

This is the accuracy part, and it applies to every DSpark target, not just K3. `_run_eval` was reducing only the loss, while each batch already computes what decides whether a draft is worth serving: per-position `accept_rate@k`, the expected accepted block length `tau`, and the confidence head's calibration against measured acceptance. Validation now accumulates those numerator/denominator sums across the split and forms each ratio once after the DP reduction, the same way the training log window does. Averaging per-batch ratios would over-weight batches with fewer measured positions, so there is a test asserting the ratio-of-sums result specifically. A diagnostic whose denominator is zero stays out of the dict instead of being logged as zero, which would read as collapsed acceptance.

After an epoch you now get `val_loss`, `val_accept_rate` and `val_tau` on the console, and `val/accept_rate@0..k`, `val/tau`, `val/confidence_abs_error`, `val/confidence_bias`, `val/confidence_cumprod_bias` in W&B.

**How to run it**

```
automodel train llm -c examples/speculative/dspark/kimi_k3_dspark.yaml
```

Point `target_model` at the K3 checkpoint and give the recipe a `validation_dataset` block so `_run_eval` has a split to score. The signal to look at is whether `val_accept_rate@k` decays smoothly with `k` and whether `val_tau` climbs during training. If `val_tau` sits near 1.0 the draft is not proposing anything useful, and if `confidence_bias` is far from 0 the confidence head is mispredicting its own acceptance.

**What I verified locally (CPU only)**

* `tests/unit_tests/speculative/`: 808 passed, 12 skipped
* `tests/unit_tests/recipes/llm/test_train_dspark_helpers.py` plus `tests/unit_tests/speculative/`: 896 passed, 12 skipped, including 5 new tests for the validation aggregation
* `ruff format` and `ruff check` clean on the changed files

**Three things I deliberately did not change, since they are performance and capacity calls rather than cleanups. Your read on these would help:**

1. `draft_kimi_k3.py:113` computes attention as explicit matmul with an fp32 softmax, materializing the full score matrix. With the example config that is roughly 7.6 GB of transient allocation per layer in the forward, and the bf16 softmax output is kept for backward, so about 5.7 GB resident for 3 layers. MiniMax M3, which is also non-causal cross plus self with no sink, uses SDPA. Note `config.py` already sets `_attn_implementation = "sdpa"` and the module never reads it. I left it alone because the fp32 softmax upcast looks deliberate and switching backends changes the precision path, so it needs a real run to confirm quality is unaffected.
2. `draft_kimi_k3.py:81` concatenates full-width hidden states per layer, about 95 MB per layer and 286 MB resident for 3 layers, where projecting first and concatenating the projections would be about 7.7 MB. GLM-5.2 does the same concatenation with a comment on the equivalence, so changing only K3 would fork the pattern.
3. `draft_kimi_k3.py:162` builds `KimiK3MLP` with the target's dense `intermediate_size=18432`, so the 3 MLP layers are roughly 1.19 B trainable parameters. GLM's draft deliberately uses `moe_intermediate_size` and notes that the drafter should stay light. A draft this heavy could eat the speculative speedup outright, and neither the YAML nor the README mentions the choice.

**Comment by khazic** [comment]
The `codecov/patch` failure on this PR is not a real coverage gap, and I would appreciate a `/ok to test` so it can clear itself.

No workflow has run on this branch's head. The previous head `2134cd03f` carried only DCO, the PR-title check, and the codecov statuses, with none of the `L0_Unit_Tests_*` or `Coverage (unit-test)` jobs that ran on #3259 and #3266. Since no report was uploaded for that commit, codecov scored the diff against an older report from an earlier push, and the line numbers no longer line up: it sees 754 statements in `train_dspark.py` where the file has 750, and every line it flags sits exactly three lines below the ones that are genuinely uncovered. It also reports nine uncovered lines in `_dspark_target_build.py`, which is at 100% on this branch.

Measured locally on the current head, over `tests/unit_tests/speculative` and `tests/unit_tests/recipes/llm` (1355 passed, 12 skipped), patch coverage is 91.1%: 163 of 179 added statements in `nemo_automodel/`. The 16 remaining lines are the recipe-assembly branches in `train_dspark.py` and `precompute_dspark_dist.py` that build a real K3 target, which the other target families do not unit-test either. `dspark/config.py` is excluded from measurement on both sides, since the repo's coverage `omit` list contains a bare `config.py` pattern that matches every file with that name.

The one real gap the check did not surface was in `draft_kimi_k3.py`, so I pushed `e2ef4f2` with tests for the two attention branches nothing reached: the `mla_use_nope=False` rejection, and the `q_lora_rank=None` path that builds a single `q_proj` instead of the Q-LoRA pair, exercised through a full forward. That file is now at 100%.

@HuiyingLi the GPU-validation request in my earlier comment still stands, and the accuracy diagnostics described there are unchanged by this push.

File: nemo_automodel/components/speculative/dspark/draft_kimi_k3.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        if self.use_output_gate:
+            self.g_proj = nn.Linear(config.hidden_size, self.num_heads * self.v_head_dim, bias=False)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        target_hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        **kwargs,
+    ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/speculative/dspark/draft_kimi_k3.py:86]: Added Google-style docstrings to the attention, decoder-layer and _forward_backbone forwards, documenting the [context | noise-block] layout.

**Comment by HuiyingLi** [comment]
GPU validation completed successfully on the full Kimi K3 target.

### Run

- Validated implementation commit: `fb010f4e73ead848525b6c85dfd7c1f9e98cf3c8`
- Recipe memory fix now on the PR branch: `84ab7566a4417a2a3f02806e9d7bc2be08525f1b`
- Slurm job: `6009830` — `COMPLETED (0:0)`, 52/52 training steps
- Topology: 64 GB200 GPUs (16 nodes x 4), FSDP2, EP64 / PP1 / CP1 / TP1
- Runtime: 2K sequence length, micro-batch size 1, 3-layer draft, HybridEP dispatcher, TE for MLA and FLA for KDA
- Peak allocated GPU memory: **128.05 GiB**
- W&B: https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/c3atr6kv
- Rank-0 run log: `/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_llm/users/huiyingl/k3/pr3263/logs/full_ep64_s2048_te_reshard_52steps_6009830_nvl72072-T01.out`

### Result

- Train loss: `3.5794` at step 1 -> `2.4879` at step 52
- Validation loss: `5.8880` after epoch 1 -> `2.4869` after epoch 13
- Validation acceptance: `0.0226` -> `0.0840`
- Validation tau: `1.0230` -> `1.0999`
- Final `val/accept_rate@0..4`: `[0.09375, 0.08181, 0.08191, 0.08132, 0.08096]`
- Final confidence absolute error: `0.09424`
- Final confidence bias: `-0.02150`
- Final confidence cumulative-product bias: `-0.00967`

Final log excerpt:

```text
step 52 | epoch 12 | loss 2.4879 | ce 5.7635 | tv 1.8371 | conf 0.2582 | accept 0.080 | tau 1.09 | lr 6.00e-05 | mem 128.05 GiB
Finished epoch 13/13 completed_steps=4 val_loss=2.4869 val_accept_rate=0.0840 val_tau=1.0999
```

The original `reshard_after_forward: false` setting OOMed during the frozen target forward at about 184 GiB. Since the target runs under `torch.no_grad()`, there is no backward pass to release the full FSDP all-gathered block parameters. Setting it to `true` reduced peak allocation to 128.05 GiB and completed the full 52-step validation. I pushed that recipe change in `84ab7566a`.

One data caveat: this run used a local 256-train / 32-validation Tulu3 slice, not responses regenerated from Kimi K3 at temperature 0.7. It validates the full GPU execution path, memory fix, distributed metric aggregation, and training behavior, but the acceptance numbers should not be treated as final PerfectBlend/serving-quality numbers.

**Review by HuiyingLi — APPROVED** [review]
